### DISCUSSION TITLE
PulpSolverError: Pulp: Error while trying to execute, use msg=True for more detailscbc

### DISCUSSION THREAD

**Discussion by marxqiu** [coin-or/pulp#384] [Q&A]
I am new to pulp, and I am trying to solve the following program using pulp:

```
from pulp import LpMaximize, LpProblem, LpStatus, lpSum, LpVariable

model = LpProblem(name="small-problem", sense=LpMaximize)

x = LpVariable(name="x", lowBound=0)
y = LpVariable(name="y", lowBound=0)

model += (2 * x + y <= 20, "red_constraint")
model += (4 * x - 5 * y >= -10, "blue_constraint")
model += (-x + 2 * y >= -2, "yellow_constraint")
model += (-x + 5 * y == 15, "green_constraint")
model += x + 2 * y

model.solve()
```

but I get the following errors:

```
---------------------------------------------------------------------------
PulpSolverError                           Traceback (most recent call last)
<ipython-input-60-667659ed42fe> in <module>
      8 model += (-x + 5 * y == 15, "green_constraint")
      9 model += x + 2 * y
---> 10 model.solve()

/opt/anaconda3/lib/python3.8/site-packages/pulp/pulp.py in solve(self, solver, **kwargs)
   1897         #time it
   1898         self.solutionTime = -clock()
-> 1899         status = solver.actualSolve(self, **kwargs)
   1900         self.solutionTime += clock()
   1901         self.restoreObjective(wasNone, dummyVar)

/opt/anaconda3/lib/python3.8/site-packages/pulp/apis/coin_api.py in actualSolve(self, lp, **kwargs)
     99     def actualSolve(self, lp, **kwargs):
    100         """Solve a well formulated lp problem"""
--> 101         return self.solve_CBC(lp, **kwargs)
    102 
    103     def available(self):

/opt/anaconda3/lib/python3.8/site-packages/pulp/apis/coin_api.py in solve_CBC(self, lp, use_mps)
    152         cbc = subprocess.Popen(args, stdout = pipe, stderr = pipe, stdin=devnull)
    153         if cbc.wait() != 0:
--> 154             raise PulpSolverError("Pulp: Error while trying to execute, use msg=True for more details" +  \
    155                                     self.path)
    156         if pipe:

PulpSolverError: Pulp: Error while trying to execute, use msg=True for more detailscbc
```

The error message is general and does not provide information on what actually broke. But when I try `model.solve(PULP_CBC_CMD(msg=1))`, it produces:

```
---------------------------------------------------------------------------
NameError                                 Traceback (most recent call last)
<ipython-input-61-0cf2e843b05c> in <module>
     12 model += x + 2 * y
     13 
---> 14 model.solve(PULP_CBC_CMD(msg=1))

NameError: name 'PULP_CBC_CMD' is not defined
```

Software:
Pulp 2.3.1(installed via conda, there is one error message I encountered when installing: `Verifying transaction: | WARNING conda.core.path_actions:verify(962): Unable to create environments file. Path not writable.` I am not sure if this is related to the problem)
Python 3.8.3
MacOS 11.1

**Comment by pchtsp** [comment]
Something you missed mentioning is that you're using python through jupyter or ipython. Your user probably has no right to execute CBC or write the log files or something. Try changing the location of the log files:
https://coin-or.github.io/pulp/guides/how_to_configure_solvers.html#configuring-where-the-cmd-solvers-write-their-temporary-files

**Reply by marxqiu** [reply]
Thanks for your quick response! I've tried to save the files in the current directory, but then I got this problem:
```
---------------------------------------------------------------------------
PulpSolverError                           Traceback (most recent call last)
<ipython-input-76-f20e97626efe> in <module>
      5 model += _var + _var2 == 1
      6 solver = pl.PULP_CBC_CMD(keepFiles=True)
----> 7 result = model.solve(solver)

/opt/anaconda3/lib/python3.8/site-packages/pulp/pulp.py in solve(self, solver, **kwargs)
   1897         #time it
   1898         self.solutionTime = -clock()
-> 1899         status = solver.actualSolve(self, **kwargs)
   1900         self.solutionTime += clock()
   1901         self.restoreObjective(wasNone, dummyVar)

/opt/anaconda3/lib/python3.8/site-packages/pulp/apis/coin_api.py in actualSolve(self, lp, callback)
    284         def actualSolve(self, lp, callback = None):
    285             """Solve a well formulated lp problem"""
--> 286             raise PulpSolverError("PULP_CBC_CMD: Not Available (check permissions on %s)" % self.pulp_cbc_path)
    287     else:
    288         def __init__(self, mip=True, msg=True, timeLimit=None,

PulpSolverError: PULP_CBC_CMD: Not Available (check permissions on /opt/anaconda3/lib/python3.8/site-packages/pulp/apis/../solverdir/cbc/osx/64/cbc)
```

**Reply by marxqiu** [reply]
I tried to pass the 'mps' file directly to cbc , but there seems no cbc:

```
-bash: /opt/anaconda3/lib/python3.8/site-packages/pulp/solverdir/cbc/osx/64/cbc: No such file or directory
```

**Reply by Winny-dev** [reply]
Hi @marxqiu, 

I had the same issue and here is how I resolved it : 
Firstly, I installed pulp package with the following code into Jupiter notebook : conda install -c conda-forge pulp. This way is absolutely not the appropriate one. 
So I decided to uninstall the package from my Conda  environment, by typing "pip uninstall pulp" in my terminal. And then I reinstall the package in using "pip install pulp" still in my terminal. 
I realised PULP_CBC_CMD  and PULP_CHOCO_CMD were available. So I think it was due to the fact I didn't install the package properly. 

<img width="619" alt="Screenshot 2021-03-10 at 12 44 51" src="https://user-images.githubusercontent.com/80397775/110632818-f7140680-819f-11eb-8908-29eef18e7e38.png">

**Reply by arunav71** [reply]
Hello @Winny-dev and @marxqiu ,

I am having a similar "solver" error for Pulp and would really appreciate if you have any suggestions.

1. I am using Pulp 2.7. I am running it through Jupyter. Both Jupyter and Pulp are installed under the "Users" directory. The snip below was my attempt at uninstalling and installing it. @Winny-dev  can you kindly confirm this is the right way to install it?

![image](https://user-images.githubusercontent.com/36560507/200859194-25ed283e-ce39-4126-bbf0-51b5d3c0d5fa.png)

2. As you can see in the snip below the error shows that Pulp is indeed in the users directory

![image](https://user-images.githubusercontent.com/36560507/200862015-33c8966e-5acb-4a2d-8bbc-741c31a4f9b5.png)

Can someone please advise? Would really appreciate the help.

Thanks,
Arunav

**Reply by arunav71** [reply]
Also, I tried running the Pulp solver test and my output is very different from the recommended screen. Do you have any suggestions on what may be going wrong? I would really appreciate the help.

Thanks,

<img width="493" alt="image" src="https://user-images.githubusercontent.com/36560507/200872144-3ecb82cb-8628-458d-ab5b-7f2ddfea4161.png">
<img width="519" alt="image" src="https://user-images.githubusercontent.com/36560507/200872322-7681bae7-eaee-414e-9063-950bbbb89c42.png">

**Reply by Sustainability4** [reply]
I am having a similar issue as well

**Reply by Pasuukki** [reply]
@arunav71 @Sustainability4 I had issue in exactly the same place (path to tmpSol does not exist). I don't know if you have similar case as me: for me it was caused by 'nan'-values in the problem, as in **something + nan*something >= nan**. These came due to faulty data and not enough validations.

@pchtsp So 'nan'-values can go through pulp to solver, which understandably does not produce a solution. Is this a useful feature in some cases, or could it be somehow fixed?

**Reply by HazemAhmad** [reply]
@arunav71 @Sustainability4 did u guys find a solution? I have the same problem but it only arises from within docker container. It works otherwise. Do u have a lead?

**Comment by peterjc** [comment]
Hitting the same error via snakemake on macOS ARM, or the Python example given.

In my case running `cbc` at the terminal is broken - probably a conda-forge packaging error [it was fixed by unistalling and reinstalling it]:

```console
❯ cbc
dyld[3804]: Library not loaded: @rpath/libgfortran.5.dylib
  Referenced from: <0B9C315B-A1DD-3527-88DB-4B90531D343F> /Users/peterjc/miniforge3/envs/pyani-plus_py312/lib/libopenblas.0.dylib
  Reason: tried: '/Users/peterjc/miniforge3/envs/pyani-plus_py312/lib/libgfortran.5.dylib' (duplicate LC_RPATH '@loader_path'), '/Users/peterjc/miniforge3/envs/pyani-plus_py312/lib/libgfortran.5.dylib' (duplicate LC_RPATH '@loader_path'), '/Users/peterjc/miniforge3/envs/pyani-plus_py312/lib/libgfortran.5.dylib' (duplicate LC_RPATH '@loader_path'), '/Users/peterjc/miniforge3/envs/pyani-plus_py312/lib/libgfortran.5.dylib' (duplicate LC_RPATH '@loader_path'), '/Users/peterjc/miniforge3/envs/pyani-plus_py312/lib/libgfortran.5.dylib' (duplicate LC_RPATH '@loader_path'), '/Users/peterjc/miniforge3/envs/pyani-plus_py312/bin/../lib/libgfortran.5.dylib' (duplicate LC_RPATH '@loader_path'), '/Users/peterjc/miniforge3/envs/pyani-plus_py312/bin/../lib/libgfortran.5.dylib' (duplicate LC_RPATH '@loader_path'), '/usr/local/lib/libgfortran.5.dylib' (no such file), '/usr/lib/libgfortran.5.dylib' (no such file, not in dyld cache)
zsh: abort      cbc
```

Note the error message is missing some white space between "details" and "cbc" (the binary name).
